src/runtime_src/driver/xclng/drm/xocl/subdev/nifd.c

```diff
@@ -0,0 +1,680 @@
+{
+    write_nifd_register(0x04, NIFD_CLK_MODES);
+    return 0;
+}
+
+static void add_breakpoint_data(unsigned int bank, unsigned int frame,
+                                unsigned int offset, unsigned int constraint)
+{
+    enum NIFD_register_offset register_offset;
+    switch (bank)
+    {
+    case 1:
+        register_offset = NIFD_CONFIG_DATA;
+        break;
+    case 2:
+        register_offset = NIFD_CONFIG_DATA_M2;
+        break;
+    default:
+        // Do not assign to either bank
+        return;
+    }
+
+    frame &= 0x3fffffff;      // Top two bits of frames must be 00
+    offset &= 0x3fffffff;     // Set the top two bits to 0 first
+    constraint &= 0x3fffffff; // Set the top two bits to 0 first
+    offset |= 0x80000000;     // Top two bits of offsets must be 10
+    constraint |= 0x40000000; // Top two bits of constraints must be 01
+
+    write_nifd_register(frame, register_offset);
+
+    // Switching to match Mahesh's test
+    write_nifd_register(constraint, register_offset);
+    write_nifd_register(offset, register_offset);
+}
+
+static long add_breakpoints_core(unsigned int *arg)
+{
+    // Format of user data:
+    //  [numBreakpoints][frameAddress][frameOffset][constraint]...[condition]
+
+    unsigned int num_breakpoints;
+    unsigned int i;
+    unsigned int frame_address;
+    unsigned int frame_offset;
+    unsigned int constraint;
+    unsigned int breakpoint_condition;
+
+    // When adding breakpoints, the clock should be stopped
+    unsigned int clock_status = (read_nifd_status() & 0x3);
+    if (clock_status != 0x3)
+        return -EINVAL;
+
+    // All breakpoints need to be set at the same time
+    clear_configuration_memory(1);
+
+    num_breakpoints = (*arg);
+
+    ++arg;
+
+    for (i = 0; i < num_breakpoints; ++i)
+    {
+        frame_address = (*arg);
+        ++arg;
+        frame_offset = (*arg);
+        ++arg;
+        constraint = (*arg);
+        ++arg;
+
+        add_breakpoint_data(1, frame_address, frame_offset, constraint);
+    }
+
+    breakpoint_condition = (*arg);
+
+    write_nifd_register(breakpoint_condition, NIFD_BREAKPOINT_CONDITION);
+
+    return 0; // Success
+}
+
+static long add_breakpoints(void __user *arg)
+{
+    // Format of user data: [numBreakpoints][frameAddress][frameOffset][constraint]...[condition]
+    unsigned int num_breakpoints;
+    unsigned int total_data_size;
+    unsigned int *kernel_memory;
+    long result;
+
+    if (copy_from_user(&num_breakpoints, arg, sizeof(unsigned int)))
+        return -EFAULT;
+
+    total_data_size = (num_breakpoints * 3 + 1 + 1) * sizeof(unsigned int);
+    kernel_memory = (unsigned int *)(vmalloc(total_data_size));
+    if (!kernel_memory)
+        return -ENOMEM;
+
+    if (copy_from_user(kernel_memory, arg, total_data_size))
+    {
+        vfree(kernel_memory);
+        return -EFAULT;
+    }
+
+    result = add_breakpoints_core(kernel_memory);
+    vfree(kernel_memory);
+
+    if (result)

```

Comment:
can we just return result?

Reply:
Yes. I changed it in the following commit.